src/coreclr/tools/superpmi/superpmi-shared/methodcontext.cpp

```diff
@@ -6683,117 +6443,174 @@ CORINFO_CLASS_HANDLE MethodContext::repGetTypeInstantiationArgument(CORINFO_CLAS
     return result;
 }
 
-void MethodContext::recAppendClassName(
-    int nBufLenIn,
-    CORINFO_CLASS_HANDLE cls,
-    bool fNamespace,
-    bool fFullInst,
-    bool fAssembly,
-    int nLenOut,
-    const char16_t* result)
-{
-    if (AppendClassName == nullptr)
-        AppendClassName = new LightWeightMap<Agnostic_AppendClassNameIn, Agnostic_AppendClassNameOut>();
-
-    // The API has two different behaviors depending on whether the input specified length is zero or non-zero:
-    // (1) zero: returns the length of the string (which the caller can use to allocate a buffer)
-    // (2) non-zero: fill as much of the buffer with the name as there is space available.
-    // We don't want the input length to be part of the key, since the caller could potentially pass in a smaller
-    // or larger buffer, and we'll fill in as much of the buffer as possible. We do need to handle both the zero
-    // and non-zero cases, though, so we'll get one record for first call using zero (with the correct buffer size result),
-    // and another for the second call with a big-enough buffer. We could presumably just store the second case
-    // (and overwrite the first record), since it contains the same output length, but it is useful to store
-    // (and see) all the JIT-EE interface calls.
-
-    Agnostic_AppendClassNameIn key;
-    ZeroMemory(&key, sizeof(key)); // Zero key including any struct padding
-    key.nBufLenIsZero = (nBufLenIn == 0) ? 1 : 0;
-    key.classHandle   = CastHandle(cls);
-    key.fNamespace    = fNamespace;
-    key.fFullInst     = fFullInst;
-    key.fAssembly     = fAssembly;
+void MethodContext::recPrint(
+    const char* name,
+    LightWeightMap<DWORDLONG, Agnostic_PrintResult>*& map,
+    DWORDLONG handle,
+    char* buffer,
+    size_t bufferSize,
+    size_t* pRequiredBufferSize,
+    size_t bytesWritten)
+{
+    if (map == nullptr)
+        map = new LightWeightMap<DWORDLONG, Agnostic_PrintResult>();
+
+    // Required size of a buffer that contains all data and null terminator.
+    UINT requiredBufferSize = UINT_MAX;
+    if (pRequiredBufferSize != nullptr)
+    {
+        requiredBufferSize = (UINT)(*pRequiredBufferSize);
+    }
+    else if (bytesWritten + 1 < bufferSize)
+    {
+        requiredBufferSize = (UINT)(bytesWritten + 1);
+    }
+
+    Agnostic_PrintResult res;
+    int index = map->GetIndex(handle);
+    if (index != -1)
+    {
+        // Merge with existing entry
+
+        res = map->GetItem(index);
+
+        if (requiredBufferSize != UINT_MAX)
+        {
+            res.requiredBufferSize = requiredBufferSize;
+        }
+
+        if (bytesWritten > res.bufferSize)
```

Comment:
```suggestion
        if (bytesWritten >= res.bufferSize)
```
bytesWritten doesn't include null character, but bufferSize does